crates/ruff_python_formatter/src/trivia.rs

```diff
@@ -9,174 +10,643 @@ use ruff_text_size::{TextLen, TextRange, TextSize};
 /// of the character, the second item the non-trivia character.
 ///
 /// Returns `None` if the range is empty or only contains trivia (whitespace or comments).
-pub(crate) fn find_first_non_trivia_character_in_range(
-    range: TextRange,
-    code: &str,
-) -> Option<(TextSize, char)> {
-    let rest = &code[range];
-    let mut char_iter = rest.chars();
-
-    while let Some(c) = char_iter.next() {
-        match c {
-            '#' => {
-                // We're now inside of a comment. Skip all content until the end of the line
-                for c in char_iter.by_ref() {
-                    if matches!(c, '\n' | '\r') {
-                        break;
-                    }
-                }
+pub(crate) fn first_non_trivia_token(offset: TextSize, code: &str) -> Option<Token> {
+    SimpleTokenizer::starts_at(offset, code)
+        .skip_trivia()
+        .next()
+}
+
+/// Returns the first non-trivia token right before `offset` or `None` if at the start of the file
+/// or all preceding tokens are trivia tokens.
+///
+/// ## Notes
+///
+/// Prefer [`first_non_trivia_token`] whenever possible because reverse lookup is expensive because of comments.
+pub(crate) fn first_non_trivia_token_rev(offset: TextSize, code: &str) -> Option<Token> {
+    SimpleTokenizer::up_to(offset, code)
+        .skip_trivia()
+        .next_back()
+}
+
+/// Returns the number of newlines between `offset` and the first non whitespace character in the source code.
+pub(crate) fn lines_before(offset: TextSize, code: &str) -> u32 {
+    let tokens = SimpleTokenizer::up_to(offset, code);
+    let mut newlines = 0u32;
+
+    for token in tokens.rev() {
```

Comment:
One of the main benefits is that we no longer repeat the same logic over and over again (and e.g. correctly handle continuation tokens)